larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,624 @@
+                    continue;
+                
+                const float distanceFromLine(clusterAssociation.GetConnectingLineDirection().GetCrossProduct(hitPosition - innerPoint).GetMagnitude());
+                
+                if (distanceFromLine > m_distanceFromLine)
+                    continue;
+                
+                extrapolatedCaloHitVector.push_back(pCaloHit);
+                caloHitToParentClusterMap[pCaloHit] = pCluster;
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TrackInEMShowerAlgorithm::IsTrackContinuous(const ClusterAssociation &clusterAssociation, CaloHitVector &extrapolatedCaloHitVector) const
+{
+    const CartesianVector &innerMergePoint(clusterAssociation.GetInnerMergePoint()), &outerMergePoint(clusterAssociation.GetOuterMergePoint());
+    const CartesianVector &trackDirection(clusterAssociation.GetConnectingLineDirection());
+    const CartesianVector trackStep(trackDirection * m_lineSegmentLength);
+
+    // to handle final segment - merge track remainder with preceding segment and if track remainder is more than half trackstep split into two
+    const float trackLength((outerMergePoint - innerMergePoint).GetMagnitude());
+    const unsigned int fullSegments(floor(trackLength / m_lineSegmentLength));
+    const float lengthOfTrackRemainder(trackLength - (fullSegments * m_lineSegmentLength));
+
+    // sort hits by projected distance from the innerMergePoint
+    std::sort(extrapolatedCaloHitVector.begin(), extrapolatedCaloHitVector.end(), SortByDistanceToLine(innerMergePoint, trackDirection));
+
+    unsigned int segmentsWithoutHits(0);
+    CaloHitVector::const_iterator caloHitIter(extrapolatedCaloHitVector.begin());
+    
+    for (unsigned int i = 0; i < (fullSegments + 1); ++i)
+    {
+        // if have run out of hits
+        if (caloHitIter == extrapolatedCaloHitVector.end())
+        {
+            ++segmentsWithoutHits;
+            if (segmentsWithoutHits > m_maxTrackGaps)
+                return false;
+            continue;
+        }
+        
+        CartesianVector lowerBoundary(innerMergePoint + (trackStep * static_cast<float>(i)));
+        CartesianVector upperBoundary(innerMergePoint + (trackStep * static_cast<float>(i + 1.f)));
+
+        if (i >= fullSegments - 1)
+        {
+            if (lengthOfTrackRemainder > m_lineSegmentLength * 0.5f)
+            {
+                lowerBoundary = lowerBoundary - (trackStep * 0.5f * (i - fullSegments + 1.f)) + (trackDirection * 0.5f * (i - fullSegments + 1.f) * lengthOfTrackRemainder);
+                upperBoundary = upperBoundary - (trackStep * 0.5f * (i - fullSegments + 2.f)) + (trackDirection * 0.5f * (i - fullSegments + 2.f) * lengthOfTrackRemainder);
```

Comment:
This looks like it is dealing with the final segment, which is not necessarily the length of a track step, but some fraction of a track step. What's the need for all of the half step adjustments here?

Reply:
So, what I do is merge the final two track segments together (i.e. a full segment + track remainder). Then, if the track remainder was originally over half of the full segment length i split the merged segments into two and loop over both of them. If the track remainder was originally under half of the full segment length i keep the merged segments as one. The half step adjustments are so that i get the correct upper and lower limits for these segments